### ISSUE TITLE
StackOverflowException in WaiterExecutor on large attempts count

### ISSUE THREAD

**Issue by flittev** [aws/aws-sdk-java-v2#3954]
### Describe the bug

`WaiterExecutor` produces `StackOverflowException` if retry attempts count is large.

### Expected Behavior

The behaviour of `WaiterExecutor` should not depend on attempts count.

### Current Behavior

StatckOverflowException stacktrace:

```
java.lang.StackOverflowError
	at java.util.Spliterator.getExactSizeIfKnown(Spliterator.java:408)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:486)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:531)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutorHelper.responseMatches(WaiterExecutorHelper.java:106)
	at java.util.Optional.map(Optional.java:215)
	at software.amazon.awssdk.utils.Either.map(Either.java:51)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutorHelper.firstWaiterAcceptorIfMatched(WaiterExecutorHelper.java:56)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.evaluate(WaiterExecutor.java:67)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.doExecute(WaiterExecutor.java:60)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.maybeRetry(WaiterExecutor.java:100)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.evaluate(WaiterExecutor.java:75)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.doExecute(WaiterExecutor.java:60)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.maybeRetry(WaiterExecutor.java:100)
...
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.evaluate(WaiterExecutor.java:75)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.doExecute(WaiterExecutor.java:60)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.maybeRetry(WaiterExecutor.java:100)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.evaluate(WaiterExecutor.java:75)
	at software.amazon.awssdk.core.internal.waiters.WaiterExecutor.doExecute(WaiterExecutor.java:60)
```

### Reproduction Steps

MRE:
```java
@Test
void largeMaxAttempts() {

    int expectedAttempts = 10_000;

    WaiterOverrideConfiguration conf =
        WaiterOverrideConfiguration.builder()
                                   .maxAttempts(expectedAttempts)
                                   .backoffStrategy(BackoffStrategy.none())
                                   .build();

    WaiterExecutor<Integer> sut =
        new WaiterExecutor<>(new WaiterConfiguration(conf),
                             Arrays.asList(
                                 WaiterAcceptor.retryOnResponseAcceptor(c -> c < expectedAttempts),
                                 WaiterAcceptor.successOnResponseAcceptor(c -> c == expectedAttempts)
                             ));

    LongAdder attemptCounter = new LongAdder();
    sut.execute(() -> {
        attemptCounter.increment();
        return attemptCounter.intValue();
    });

    Assert.assertEquals(attemptCounter.intValue(), expectedAttempts);
}
```

### AWS Java SDK version used

2.20.53

### JDK version used

openjdk version "1.8.0_292"

### Operating System and version

all

**Comment by debora-ito** [comment]
@flittev Even though I understand the issue, setting 10,000 max attempts is a big jump from the default of 3, we definitely would not recommend it, regardless of the StackOverflowError. 

Do you have a use case for such a large max retry attempt number?

**Comment by flittev** [comment]
@debora-ito The exception is thrown much earlier than 10,000 (depends on JVM configuration I think, on my local machine it's just around 1,200). Also there is no any limitations/recommendations on the interface/class usage mentioned in documentation.

We didn't face `StackOverflowException` in our case though we have got a suspiciously large stacktrace for our configuration. That's why the issue was submitted.

We have a logic configuring max attempts for `DefaultWaiter` based on the `job_timeout = 45 minutes` and `FixedDelayBackoffStrategy` with `duration = 30 seconds`. So the `maxAttempts` are calculated to `45m/30s=90`. Also 90 is much lower than 1200, it could be reached with easy by `job_time=2h` and default `duration=5s`.

**Comment by debora-ito** [comment]
The PR got merged and was released in SDK version `2.20.88`. Thank you!
